File: crates/defi-protocols/src/yield_source/generic_yield.rs
use alloy::primitives::{Address, U256};
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::YieldSource;
use defi_core::types::*;

pub struct GenericYield {
    name: String,
    interface: String,
}

impl GenericYield {
    pub fn from_entry(name: String, interface: String) -> Self {
        Self { name, interface }
    }
}

#[async_trait]
impl YieldSource for GenericYield {
    fn name(&self) -> &str {
        &self.name
    }

    async fn get_yields(&self) -> Result<Vec<YieldInfo>> {
        Err(DefiError::Unsupported(format!(
            "[{}] get_yields requires RPC",
            self.name
        )))
    }

    async fn build_deposit(
        &self,
        _pool: &str,
        _amount: U256,
        _recipient: Address,
    ) -> Result<DeFiTx> {
        Err(DefiError::Unsupported(format!(
            "[{}] Yield interface '{}' requires a protocol-specific adapter. \
            Supported: 'pendle_v2' (Pendle). Protocols like Steer (managed liquidity), \
            Liminal (yield optimization), and Altura (gaming yield) need custom deposit logic.",
            self.name, self.interface
        )))
    }

    async fn build_withdraw(
        &self,
        _pool: &str,
        _amount: U256,
        _recipient: Address,
    ) -> Result<DeFiTx> {
        Err(DefiError::Unsupported(format!(
            "[{}] Yield interface '{}' requires a protocol-specific adapter. \
            Supported: 'pendle_v2' (Pendle). Protocols like Steer (managed liquidity), \
            Liminal (yield optimization), and Altura (gaming yield) need custom withdraw logic.",
            self.name, self.interface
        )))
    }
}

File: ts/packages/defi-protocols/src/yield_source/generic_yield.ts
import type { Address } from "viem";
import type { IYieldSource } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type YieldInfo,
  type DeFiTx,
} from "@hypurrquant/defi-core";

export class GenericYieldAdapter implements IYieldSource {
  private readonly protocolName: string;
  private readonly interfaceName: string;

  constructor(entry: ProtocolEntry, _rpcUrl?: string) {
    this.protocolName = entry.name;
    this.interfaceName = entry.interface;
  }

  name(): string {
    return this.protocolName;
  }

  async getYields(): Promise<YieldInfo[]> {
    throw DefiError.unsupported(`[${this.protocolName}] getYields requires RPC`);
  }

  async buildDeposit(_pool: string, _amount: bigint, _recipient: Address): Promise<DeFiTx> {
    throw DefiError.unsupported(
      `[${this.protocolName}] Yield interface '${this.interfaceName}' requires a protocol-specific adapter. ` +
      `Supported: 'pendle_v2' (Pendle). Protocols like Steer (managed liquidity), ` +
      `Liminal (yield optimization), and Altura (gaming yield) need custom deposit logic.`,
    );
  }

  async buildWithdraw(_pool: string, _amount: bigint, _recipient: Address): Promise<DeFiTx> {
    throw DefiError.unsupported(
      `[${this.protocolName}] Yield interface '${this.interfaceName}' requires a protocol-specific adapter. ` +
      `Supported: 'pendle_v2' (Pendle). Protocols like Steer (managed liquidity), ` +
      `Liminal (yield optimization), and Altura (gaming yield) need custom withdraw logic.`,
    );
  }
}
